File: agenkit/adapters/llm/ollama.py
"""
Ollama adapter for Agenkit.

This module provides an adapter for Ollama, the popular local LLM tool.
Supports running models locally for development, testing, and on-premises deployments.
"""

from collections.abc import AsyncIterator
from typing import Any

from agenkit.adapters.llm.base import LLM
from agenkit.interfaces import Message

try:
    from ollama import AsyncClient
except ImportError as e:
    raise ImportError(
        "The ollama package is required to use OllamaLLM. "
        "Install it with: pip install ollama>=0.1.0"
    ) from e


class OllamaLLM(LLM):
    """
    Ollama adapter for local LLM inference.

    Wraps the Ollama Python SDK to provide local model access through the
    Agenkit interface. Ideal for development, testing, and on-premises deployments.

    Args:
        model: Model identifier (e.g., "llama2", "mistral", "codellama")
        base_url: Ollama server URL (default: http://localhost:11434)
        **client_kwargs: Additional arguments passed to AsyncClient

    Example:
        >>> from agenkit.adapters.llm import OllamaLLM
        >>> from agenkit import Message
        >>>
        >>> # Basic usage (assumes Ollama running locally)
        >>> llm = OllamaLLM(model="llama2")
        >>> messages = [Message(role="user", content="Hello!")]
        >>> response = await llm.complete(messages)
        >>> print(response.content)

    Remote Ollama server:
        >>> llm = OllamaLLM(
        ...     model="mistral",
        ...     base_url="http://ollama-server:11434"
        ... )

    Streaming example:
        >>> async for chunk in llm.stream(messages):
        ...     print(chunk.content, end="", flush=True)

    Popular models:
        - llama2 - Meta's Llama 2 (7B, 13B, 70B)
        - llama3 - Meta's Llama 3
        - mistral - Mistral 7B
        - codellama - Code-focused Llama
        - phi - Microsoft's small efficient model
        - gemma - Google's open model
        - qwen - Alibaba's Qwen
        - neural-chat - Intel's fine-tuned model

    Setup:
        1. Install Ollama: https://ollama.ai/download
        2. Pull a model: `ollama pull llama2`
        3. Verify: `ollama list`
        4. Use with Agenkit!
    """

    def __init__(
        self,
        model: str = "llama2",
        base_url: str = "http://localhost:11434",
        **client_kwargs: Any,
    ):
        """Initialize Ollama LLM adapter."""
        self._model = model
        self._client = AsyncClient(host=base_url, **client_kwargs)

    @property
    def model(self) -> str:
        """Return the model identifier."""
        return self._model

    async def complete(
        self,
        messages: list[Message],
        temperature: float = 1.0,
        max_tokens: int | None = None,
        **kwargs: Any,
    ) -> Message:
        """
        Generate a completion from Ollama.

        Args:
            messages: Conversation history as Agenkit Messages
            temperature: Sampling temperature (0.0-2.0)
            max_tokens: Maximum tokens to generate (None for model default)
            **kwargs: Additional Ollama options (top_p, top_k, etc.)

        Returns:
            Response as Agenkit Message with metadata including:
            - model: Model used
            - eval_count: Number of tokens in response
            - total_duration: Total inference time (ns)

        Example:
            >>> messages = [
            ...     Message(role="system", content="You are a helpful assistant."),
            ...     Message(role="user", content="What is 2+2?")
            ... ]
            >>> response = await llm.complete(messages, temperature=0.2)
            >>> print(response.content)
        """
        # Convert Agenkit Messages to Ollama format
        ollama_messages = self._convert_messages(messages)

        # Build options
        options: dict[str, Any] = {
            "temperature": temperature,
        }

        # Add num_predict (Ollama's max_tokens)
        if max_tokens is not None:
            options["num_predict"] = max_tokens

        # Merge additional kwargs
        options.update(kwargs)

        # Call Ollama API
        response = await self._client.chat(
            model=self._model,
            messages=ollama_messages,
            options=options,
        )

        # Build metadata
        metadata: dict[str, Any] = {
            "model": response.get("model", self._model),
        }

        # Add performance metrics if available
        if "eval_count" in response:
            metadata["eval_count"] = response["eval_count"]
        if "total_duration" in response:
            metadata["total_duration"] = response["total_duration"]
        if "load_duration" in response:
            metadata["load_duration"] = response["load_duration"]

        # Extract message content
        message_content = response.get("message", {})
        content = message_content.get("content", "")

        # Convert response to Agenkit Message
        return Message(
            role="agent",
            content=content,
            metadata=metadata,
        )

    async def stream(
        self,
        messages: list[Message],
        temperature: float = 1.0,
        max_tokens: int | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[Message]:
        """
        Stream completion chunks from Ollama.

        Args:
            messages: Conversation history as Agenkit Messages
            temperature: Sampling temperature (0.0-2.0)
            max_tokens: Maximum tokens to generate
            **kwargs: Additional Ollama options

        Yields:
            Message chunks as they arrive from Ollama

        Example:
            >>> messages = [Message(role="user", content="Count to 10")]
            >>> async for chunk in llm.stream(messages):
            ...     print(chunk.content, end="", flush=True)
        """
        # Convert messages
        ollama_messages = self._convert_messages(messages)

        # Build options
        options: dict[str, Any] = {
            "temperature": temperature,
        }

        if max_tokens is not None:
            options["num_predict"] = max_tokens

        options.update(kwargs)

        # Stream from Ollama API
        async for chunk in await self._client.chat(
            model=self._model,
            messages=ollama_messages,
            options=options,
            stream=True,
        ):
            # Extract content from chunk
            message_content = chunk.get("message", {})
            content = message_content.get("content", "")

            if content:
                yield Message(
                    role="agent",
                    content=content,
                    metadata={"streaming": True, "model": self._model},
                )

    def _convert_messages(self, messages: list[Message]) -> list[dict[str, str]]:
        """
        Convert Agenkit Messages to Ollama format.

        Ollama expects:
        - role: "system", "user", or "assistant"
        - content: str
        """
        ollama_messages = []

        for msg in messages:
            # Map roles
            if msg.role in ("system", "user"):
                role = msg.role
            else:
                # Map "agent" and others to "assistant"
                role = "assistant"

            ollama_messages.append({"role": role, "content": str(msg.content)})

        return ollama_messages

    def unwrap(self) -> AsyncClient:
        """
        Get the underlying Ollama AsyncClient.

        Returns:
            The AsyncClient for direct API access

        Example:
            >>> llm = OllamaLLM()
            >>> client = llm.unwrap()
            >>> # Use Ollama-specific features
            >>> models = await client.list()
            >>> await client.pull("mistral")
        """
        return self._client

File: agenkit-ts/src/adapters/ollama.ts
/**
 * Ollama LLM adapter for local model inference.
 *
 * Provides integration with Ollama for running local LLMs like Llama 2, Mistral,
 * CodeLlama, and other models. Ideal for development, testing, and on-premises
 * deployments without API costs.
 *
 * @example
 * ```typescript
 * import { OllamaAdapter } from './adapters/ollama';
 *
 * const adapter = new OllamaAdapter({
 *   model: 'llama2',
 *   baseURL: 'http://localhost:11434',
 *   temperature: 0.7,
 * });
 *
 * const response = await adapter.process({
 *   role: 'user',
 *   content: 'Hello, Llama!',
 * });
 * console.log(response.content);
 * ```
 *
 * Setup:
 * 1. Install Ollama: https://ollama.ai/download
 * 2. Pull a model: `ollama pull llama2`
 * 3. Verify: `ollama list`
 * 4. Use with AgentKit!
 */

import { Agent, Message, createMessage, validateMessage } from '../core/interfaces';

/**
 * Configuration for Ollama adapter.
 */
export interface OllamaConfig {
  /** Model to use (e.g., 'llama2', 'mistral', 'codellama') */
  model?: string;

  /** Ollama server base URL */
  baseURL?: string;

  /** Temperature for sampling (0.0 - 2.0) */
  temperature?: number;

  /** Maximum tokens to generate */
  maxTokens?: number;

  /** Top-p sampling parameter */
  topP?: number;

  /** Top-k sampling parameter */
  topK?: number;

  /** Request timeout in milliseconds */
  timeout?: number;
}

/**
 * Message format for Ollama API.
 */
interface OllamaMessage {
  role: string;
  content: string;
}

/**
 * Request format for Ollama chat API.
 */
interface OllamaChatRequest {
  model: string;
  messages: OllamaMessage[];
  stream: boolean;
  options?: {
    temperature?: number;
    num_predict?: number;
    top_p?: number;
    top_k?: number;
  };
}

/**
 * Response format from Ollama chat API.
 */
interface OllamaChatResponse {
  model: string;
  created_at: string;
  message: {
    role: string;
    content: string;
  };
  done: boolean;
  total_duration?: number;
  load_duration?: number;
  prompt_eval_count?: number;
  prompt_eval_duration?: number;
  eval_count?: number;
  eval_duration?: number;
}

/**
 * Ollama adapter implementing the Agent interface.
 *
 * Features:
 * - Support for Llama 2, Mistral, CodeLlama, and other Ollama models
 * - Local inference (no API key required)
 * - Zero cost for development and testing
 * - Privacy-preserving (data never leaves your machine)
 * - Configurable parameters (temperature, max_tokens, etc.)
 * - Automatic message format conversion
 */
export class OllamaAdapter implements Agent {
  private config: Required<OllamaConfig>;

  /**
   * Creates a new Ollama adapter.
   *
   * @param config - Configuration options
   */
  constructor(config: OllamaConfig = {}) {
    this.config = {
      model: config.model || 'llama2',
      baseURL: config.baseURL || 'http://localhost:11434',
      temperature: config.temperature ?? 0.7,
      maxTokens: config.maxTokens ?? 1024,
      topP: config.topP ?? 1.0,
      topK: config.topK ?? 40,
      timeout: config.timeout ?? 120000, // 2 minutes for local inference
    };
  }

  /**
   * Agent name for identification.
   */
  get name(): string {
    return `ollama-${this.config.model}`;
  }

  /**
   * Agent capabilities.
   */
  get capabilities(): string[] {
    return ['completion', 'local-inference', 'no-api-key'];
  }

  /**
   * Process a message through Ollama.
   *
   * @param message - Input message
   * @returns Response message with content and metadata
   */
  async process(message: Message): Promise<Message> {
    validateMessage(message);

    // Convert single message to array for API
    const messages: OllamaMessage[] = [{
      role: message.role,
      content: message.content,
    }];

    // Build request
    const request: OllamaChatRequest = {
      model: this.config.model,
      messages,
      stream: false,
      options: {
        temperature: this.config.temperature,
        num_predict: this.config.maxTokens,
        top_p: this.config.topP,
        top_k: this.config.topK,
      },
    };

    try {
      // Make HTTP request to Ollama
      const response = await fetch(`${this.config.baseURL}/api/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(this.config.timeout),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Ollama API error (${response.status}): ${errorText}`);
      }

      const data: OllamaChatResponse = await response.json();

      // Build metadata
      const metadata: Record<string, any> = {
        model: data.model,
        created_at: data.created_at,
      };

      if (data.total_duration) {
        metadata.total_duration_ms = data.total_duration / 1_000_000;
      }

      if (data.prompt_eval_count || data.eval_count) {
        metadata.usage = {
          prompt_tokens: data.prompt_eval_count || 0,
          completion_tokens: data.eval_count || 0,
          total_tokens: (data.prompt_eval_count || 0) + (data.eval_count || 0),
        };
      }

      // Create response message
      return createMessage({
        role: data.message.role,
        content: data.message.content,
        metadata,
      });
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new Error(`Ollama request timeout after ${this.config.timeout}ms`);
        }
        throw new Error(`Ollama request failed: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Stream a response from Ollama (not yet implemented).
   *
   * @param message - Input message
   * @returns Async iterator of response chunks
   */
  async *stream(message: Message): AsyncGenerator<Message, void, unknown> {
    validateMessage(message);

    // Convert single message to array for API
    const messages: OllamaMessage[] = [{
      role: message.role,
      content: message.content,
    }];

    // Build request
    const request: OllamaChatRequest = {
      model: this.config.model,
      messages,
      stream: true,
      options: {
        temperature: this.config.temperature,
        num_predict: this.config.maxTokens,
        top_p: this.config.topP,
        top_k: this.config.topK,
      },
    };

    try {
      // Make HTTP request to Ollama
      const response = await fetch(`${this.config.baseURL}/api/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(this.config.timeout),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Ollama API error (${response.status}): ${errorText}`);
      }

      if (!response.body) {
        throw new Error('Response body is null');
      }

      // Parse streaming response
      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();

        if (done) break;

        buffer += decoder.decode(value, { stream: true });

        // Split by newlines to handle multiple JSON objects
        const lines = buffer.split('\n');
        buffer = lines.pop() || ''; // Keep last incomplete line in buffer

        for (const line of lines) {
          if (line.trim()) {
            try {
              const data: OllamaChatResponse = JSON.parse(line);

              // Yield chunk
              yield createMessage({
                role: data.message.role,
                content: data.message.content,
                metadata: {
                  model: data.model,
                  done: data.done,
                },
              });

              if (data.done) {
                return;
              }
            } catch (e) {
              // Skip malformed JSON lines
              continue;
            }
          }
        }
      }
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new Error(`Ollama stream timeout after ${this.config.timeout}ms`);
        }
        throw new Error(`Ollama stream failed: ${error.message}`);
      }
      throw error;
    }
  }
}
